### PR TITLE
sql: export operator and func info

### PR DESCRIPTION
**Pull request by madelynnblue**
Needed so we can autogenerate documentation.

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/5875)

### CODE REVIEW

File: sql/parser/type_check.go [hunk 1]
```diff
@@ -25,14 +25,22 @@ import (
...
 var (
-	typeBytes     = func(MapArgs, DTuple) (Datum, error) { return DummyBytes, nil }
-	typeDate      = func(MapArgs, DTuple) (Datum, error) { return DummyDate, nil }
-	typeFloat     = func(MapArgs, DTuple) (Datum, error) { return DummyFloat, nil }
-	typeDecimal   = func(MapArgs, DTuple) (Datum, error) { return DummyDecimal, nil }
-	typeInt       = func(MapArgs, DTuple) (Datum, error) { return DummyInt, nil }
-	typeInterval  = func(MapArgs, DTuple) (Datum, error) { return DummyInterval, nil }
-	typeString    = func(MapArgs, DTuple) (Datum, error) { return DummyString, nil }
-	typeTimestamp = func(MapArgs, DTuple) (Datum, error) { return DummyTimestamp, nil }
+	// TypeBytes returns a bytes datum.
```
Comment (tamird) [inline sql/parser/type_check.go]: these need to be exported?

File: sql/parser/builtins.go [hunk 2]
```diff
@@ -55,13 +55,14 @@ var (
 	errLogOfZero         = errors.New("cannot take logarithm of zero")
 )
 
-type argTypes []reflect.Type
+// ArgTypes accepts a specific number of type of arguments.
```
Comment (tamird) [inline sql/parser/builtins.go]: "...number of argument types"?

File: sql/parser/builtins.go [hunk 3]
```diff
@@ -73,30 +74,32 @@ func (a argTypes) match(types argTypes) bool {
...
 			return false
 		}
 	}
 	return true
 }
 
-type builtin struct {
-	types      typeList
-	returnType func(MapArgs, DTuple) (Datum, error)
+// Builtin contains information for a builtin function.
```
Comment (tamird) [inline sql/parser/builtins.go]: "is a built-in function"?

File: sql/parser/builtins.go [hunk 4]
```diff
@@ -106,39 +109,39 @@ type builtin struct {
 	fn     func(EvalContext, DTuple) (Datum, error)
 }
 
-var timestampMinusBinOp = binOps[binArgs{Minus, timestampType, timestampType}]
+var timestampMinusBinOp = BinOps[BinArgs{Minus, timestampType, timestampType}]
 
-// The map from function name to function data. Keep the list of functions
-// sorted please.
-var builtins = map[string][]builtin{
+// Builtins map from function name to function data. Keep the list of
```
Comment (tamird) [inline sql/parser/builtins.go]: I'd remove the bit about sortedness in a doc comment. Also "Builtins contains the built-in functions indexed by name"?

File: sql/parser/eval.go [hunk 5]
```diff
@@ -46,52 +46,54 @@ var (
...
+// UnaryOp contains unary return type and implementation.
+type UnaryOp struct {
+	ReturnType Datum
 	fn         func(EvalContext, Datum) (Datum, error)
 }
 
-type unaryArgs struct {
-	op      UnaryOp
-	argType reflect.Type
+// UnaryArgs contains unary operations and types.
```
Comment (tamird) [inline sql/parser/eval.go]: is a unary operation

File: sql/parser/eval.go [hunk 6]
```diff
@@ -46,52 +46,54 @@ var (
 // secondsInDay is the number of seconds in a day.
 const secondsInDay = 24 * 60 * 60
 
-type unaryOp struct {
-	returnType Datum
+// UnaryOp contains unary return type and implementation.
```
Comment (tamird) [inline sql/parser/eval.go]: "is a unary operator"?

File: sql/parser/eval.go [hunk 7]
```diff
@@ -100,44 +102,46 @@ var unaryOps = map[unaryArgs]unaryOp{
...
+type BinOp struct {
+	ReturnType Datum
 	fn         func(EvalContext, Datum, Datum) (Datum, error)
 }
 
-type binArgs struct {
-	op        BinaryOp
-	leftType  reflect.Type
-	rightType reflect.Type
+// BinArgs contains binary operations and types.
```
Comment (tamird) [inline sql/parser/eval.go]: is a binary operation

File: sql/parser/eval.go [hunk 8]
```diff
@@ -100,44 +102,46 @@ var unaryOps = map[unaryArgs]unaryOp{
...
+		ReturnType: DummyInt,
 		fn: func(_ EvalContext, d Datum) (Datum, error) {
 			return ^d.(DInt), nil
 		},
 	},
 }
 
-type binOp struct {
-	returnType Datum
+// BinOp contains binary return type and implementation.
```
Comment (tamird) [inline sql/parser/eval.go]: is a binary operator

File: sql/parser/eval.go [hunk 9]
```diff
@@ -356,308 +360,310 @@ var binOps = map[binArgs]binOp{
...
-	rightType reflect.Type
+// CmpArgs contains comparison operations and types.
+type CmpArgs struct {
+	Op        ComparisonOp
+	LeftType  reflect.Type
+	RightType reflect.Type
 }
 
-type cmpOp struct {
+// CmpOp contains comparison implementation.
```
Comment (tamird) [inline sql/parser/eval.go]: is a comparison operator

File: sql/parser/eval.go [hunk 10]
```diff
@@ -356,308 +360,310 @@ var binOps = map[binArgs]binOp{
...
 			return left.(DInt) >> uint(right.(DInt)), nil
 		},
 	},
 }
 
-type cmpArgs struct {
-	op        ComparisonOp
-	leftType  reflect.Type
-	rightType reflect.Type
+// CmpArgs contains comparison operations and types.
```
Comment (tamird) [inline sql/parser/eval.go]: is a comparison operation

**Comment by mjibson** [comment]
Review status: 0 of 7 files reviewed at latest revision, 10 unresolved discussions.

---

*[sql/parser/type_check.go, line 28 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5875#-KEbr6OWoEqsJi1RB9dH-r1-28:-KEbzOqt79jHP0kynX-9:-1169615491) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cca6809e9261853b7661014dd8a93fbff341ec3/sql/parser/type_check.go#L28)):*
Yes. Since `ReturnType` is a function that returns a datum, not just a datum, we need to compare the function pointers and map those to a string. See https://github.com/cockroachdb/docs/blob/64265d14a7e23698fa92cfe4da81956e168fe003/generate/funcs.go#L114

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5875)*

**Comment by mjibson** [comment]
Review status: 0 of 7 files reviewed at latest revision, 10 unresolved discussions.

---

*[sql/parser/builtins.go, line 58 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5875#-KEbr6OWoEqsJi1RB9dB-r1-58:-KEc-DEQEVozspovwW1s:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cca6809e9261853b7661014dd8a93fbff341ec3/sql/parser/builtins.go#L58)):*
Done.

---

*[sql/parser/builtins.go, line 99 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5875#-KEbr6OWoEqsJi1RB9dB-r1-99:-KEc-G7_ksZKrVep5zKB:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cca6809e9261853b7661014dd8a93fbff341ec3/sql/parser/builtins.go#L99)):*
Done.

---

*[sql/parser/builtins.go, line 114 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5875#-KEbr6OWoEqsJi1RB9dB-r1-114:-KEc-PwgDUIBGh-cKN74:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cca6809e9261853b7661014dd8a93fbff341ec3/sql/parser/builtins.go#L114)):*
Done.

---

*[sql/parser/eval.go, line 49 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5875#-KEbr6OWoEqsJi1RB9dD-r1-49:-KEc-W2O4iXaNRh26L0c:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cca6809e9261853b7661014dd8a93fbff341ec3/sql/parser/eval.go#L49)):*
Done.

---

*[sql/parser/eval.go, line 55 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5875#-KEbr6OWoEqsJi1RB9dD-r1-55:-KEc-YE5wCyDT-5afTZ5:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cca6809e9261853b7661014dd8a93fbff341ec3/sql/parser/eval.go#L55)):*
Done.

---

*[sql/parser/eval.go, line 113 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5875#-KEbr6OWoEqsJi1RB9dD-r1-113:-KEc-bkKEtmTalohJ00A:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cca6809e9261853b7661014dd8a93fbff341ec3/sql/parser/eval.go#L113)):*
Done.

---

*[sql/parser/eval.go, line 119 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5875#-KEbr6OWoEqsJi1RB9dD-r1-119:-KEc-bVtFAkH6DvuzkRx:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cca6809e9261853b7661014dd8a93fbff341ec3/sql/parser/eval.go#L119)):*
Done.

---

*[sql/parser/eval.go, line 391 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5875#-KEbr6OWoEqsJi1RB9dD-r1-391:-KEc-gMxMjQQJvYXNaLh:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cca6809e9261853b7661014dd8a93fbff341ec3/sql/parser/eval.go#L391)):*
Done.

---

*[sql/parser/eval.go, line 398 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5875#-KEbr6OWoEqsJi1RB9dD-r1-398:-KEc-g8fA_PB8uphL8zf:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cca6809e9261853b7661014dd8a93fbff341ec3/sql/parser/eval.go#L398)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5875)*

**Comment by tamird** [comment]
:lgtm:

---

Reviewed 5 of 7 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5875#-:-KEc2jLxhtWNd799rskY:1417088721)*

**Pull request merged**